storage: instrument eager ReplicaGC

tbg opened the pull request:
We have been seeing long startup times which disappear spontaneously. During a
restart of the beta cluster, the following goroutine was observed, which suggests
that we were spending a lot of time GCing replicas on startup.

```
engine              ??:0                     _Cfunc_DBIterNext(#324, #323, 0, 0, 0, 0, 0, 0, 0)
engine              rocksdb.go:1135          (*rocksDBIterator).Next(#235)
storage             replica_data_iter.go:104 (*ReplicaDataIterator).Next(#316)
storage             store.go:1748            (*Store).destroyReplicaData(#109, #317, 0, 0)
storage             store.go:841             (*Store).Start.func2(0x101b, #300, 0x36, 0x40, #301, 0x36, 0x40, #315, 0x3, 0x4, ...)
storage             store.go:734             IterateRangeDescriptors.func1(#306, 0x40, 0x41, #307, 0x92, 0x92, #341, 0, 0x186c, 0x4000, ...)
engine              mvcc.go:1593             MVCCIterate(#329, #68, #47, #81, #232, 0x9, 0x10, #233, 0xb, 0x10, ...)
storage             store.go:738             IterateRangeDescriptors(#330, #196, #47, #81, #195, #179, #110)
storage             store.go:867             (*Store).Start(#109, #330, #196, #179, #185, 0x1)
server              node.go:405              (*Node).initStores(#78, #330, #196, #98, 0x1, 0x1, #179, 0, #55)
server              node.go:330              (*Node).start(#78, #330, #196, #42, #129, #98, 0x1, 0x1, 0, 0, ...)
server              server.go:431            (*Server).Start(#5, #330, #196, #95, 0x1)
cli                 start.go:368             runStart(#34, #178, 0, 0x9, 0, 0)
cobra               command.go:599           (*Command).execute(#34, #177, 0x9, 0x9, #34, #177)
cobra               command.go:689           (*Command).ExecuteC(#33, #70, #343, #72)
cobra               command.go:648           (*Command).Execute(#33, #71, #343)
cli                 cli.go:96                Run(#64, 0xa, 0xa, 0, 0)
main                main.go:37               main()
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8985)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 841 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8985#-KQWULQ2ZPQrhD7It7cy:-KQWULQ2ZPQrhD7It7cz:b-25agx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ec548a549c8ad2325accfd4fd77c51affc0af2/storage/store.go#L841)):*
> ```Go
> 				// to avoid creating a new replica without a valid replica ID
> 				// (which is necessary to have a non-nil raft group)
> 				log.Infof(ctx, "eagerly GC'ed %+v", desc)
> ```

The action hasn't been performed yet so the past tense is a bit off. How about `destroying local data: %+v`? This matches up with the log message in `replica_gc_queue.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8985#-:-KQWUfpiHireB5TWJteo:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store.go, line 841 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8985#-KQWULQ2ZPQrhD7It7cy:-KQWVnB-paM47D5tFyiS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ec548a549c8ad2325accfd4fd77c51affc0af2/storage/store.go#L841)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The action hasn't been performed yet so the past tense is a bit off. How about `destroying local data: %+v`? This matches up with the log message in `replica_gc_queue.go`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8985)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: instrument eager ReplicaGC

We have been seeing long startup times which disappear spontaneously. During a
restart of the beta cluster, the following goroutine was observed, which suggests
that we were spending a lot of time GCing replicas on startup.

    engine              ??:0                     _Cfunc_DBIterNext(#324, #323, 0, 0, 0, 0, 0, 0, 0)
    engine              rocksdb.go:1135          (*rocksDBIterator).Next(#235)
    storage             replica_data_iter.go:104 (*ReplicaDataIterator).Next(#316)
    storage             store.go:1748            (*Store).destroyReplicaData(#109, #317, 0, 0)
    storage             store.go:841             (*Store).Start.func2(0x101b, #300, 0x36, 0x40, #301, 0x36, 0x40, #315, 0x3, 0x4, ...)
    storage             store.go:734             IterateRangeDescriptors.func1(#306, 0x40, 0x41, #307, 0x92, 0x92, #341, 0, 0x186c, 0x4000, ...)
    engine              mvcc.go:1593             MVCCIterate(#329, #68, #47, #81, #232, 0x9, 0x10, #233, 0xb, 0x10, ...)
    storage             store.go:738             IterateRangeDescriptors(#330, #196, #47, #81, #195, #179, #110)
    storage             store.go:867             (*Store).Start(#109, #330, #196, #179, #185, 0x1)
    server              node.go:405              (*Node).initStores(#78, #330, #196, #98, 0x1, 0x1, #179, 0, #55)
    server              node.go:330              (*Node).start(#78, #330, #196, #42, #129, #98, 0x1, 0x1, 0, 0, ...)
    server              server.go:431            (*Server).Start(#5, #330, #196, #95, 0x1)
    cli                 start.go:368             runStart(#34, #178, 0, 0x9, 0, 0)
    cobra               command.go:599           (*Command).execute(#34, #177, 0x9, 0x9, #34, #177)
    cobra               command.go:689           (*Command).ExecuteC(#33, #70, #343, #72)
    cobra               command.go:648           (*Command).Execute(#33, #71, #343)
    cli                 cli.go:96                Run(#64, 0xa, 0xa, 0, 0)
    main                main.go:37               main()

tamird commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8985)*
<!-- Sent from Reviewable.io -->
